### PR TITLE
Adding support for GDScript

### PR DESCRIPTION
**Pull request by ofrank123**
Adding support for GDScript.  Importantly, this will only work if [the PR](https://github.com/godotengine/godot/pull/35864) to change over the godot LSP to TCP is accepted.

### CODE REVIEW

**Comment by ofrank123** [comment]
can't tell why the travis is failing, it says it has something to do with the header.

**Comment by danielmartin** [comment]
@ofrank The build failure is because you removed a semicolon from the header of `lsp-mode.el` (https://github.com/emacs-lsp/lsp-mode/pull/1377/files#diff-b1e9bc82c3b6bd6e95470e18ee41a19bR1). You need to revert that change. See https://www.gnu.org/software/emacs/manual/html_node/elisp/Library-Headers.html for the expected library header format of Emacs packages.

File: lsp-clients.el [hunk 1]
```diff
@@ -28,6 +28,30 @@
 (require 'dash)
 (require 'dash-functional)
 (require 'rx)
+(require 'lsp-pyls)
```
Comment (yyoncho) [inline lsp-clients.el]: revert that, use  `lsp-client-packages` to import the new library.

File: lsp-gdscript.el [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  :package-version '(lsp-mode . "6.1"))
+
+(defcustom lsp-gdscript-port 6008
+  "Port to connect server to"
+  :type 'integer
+  :group 'lsp-gdscript)
+
+(defun lsp-gdscript-tcp-connect-to-port ()
+  (list
+   :connect (lambda (filter sentinel name environment-fn)
```
Comment (yyoncho) [inline lsp-gdscript.el]: 1. you should set the sentinel to the tcp connection. Test if it works by killing the gdscript server.
2. Prefix environment-fn with _ to fix travis.

**Comment by ofrank123** [comment]
Changes requested have been applied, and I tested the sentinel so if the engine closes while the client is still connected it successfully disconnects.

**Comment by yyoncho** [comment]
Are we ready to merge that? If yes, before me merging the pr can you:

1. Can you remove the WIP from the title?
2. Can you update the readme with supported servers and eventually create a wiki/blogpost/etc with details on how to configure the server(and so on).

I am fine with the change as it is from elisp perspective. I will leave it open if someone (cc @NathanLovato @Thorhian @howdoicomputer ) wants to test it.

**Comment by NathanLovato** [comment]
I tested the tool already before with a custom engine build, it was working well. Godot is in an unstable state now, I'll see if they cherry pick and adapt the code to the stable branch or if it's testable right now on the engine's side.

**Comment by ofrank123** [comment]
Yes, ready for merge.  Anyone looking to test should cherry pick my commit to master on godot into the 3.2 branch and build that.  I will update the wiki when I have time.

**Comment by NathanLovato** [comment]
> Anyone looking to test should cherry pick my commit to master on godot into the 3.2 branch and build that.

Unfortunately, right now, it leads to a merge conflict. But with a bit of C++ fu you can test it doing so.

**Comment by yyoncho** [comment]
Thank you for contributing to `lsp-mode`!

**Comment by Narice** [comment]
Hey @NathanLovato, how did you make it work on emacs?
I built successfully Godot with the command you gave on the TCP LSP pull request on Godot's Github, and have lsp-mode installed correctly (up-to-date doom emacs) but it doesn't work, did you do something special to make it work? (I'm pretty new to emacs, sorry if it is quite trivial)

**Comment by yyoncho** [comment]
> but it doesn't work, 

What do you see?

**Comment by Narice** [comment]
There is no syntax highlighting and running the command 'lsp' gives:
`LSP :: No LSP server for fundamental-mode(check *lsp-log*).`

**Comment by Narice** [comment]
There is no lsp-log buffer so I don't know where to find more information about the issue

**Comment by yyoncho** [comment]
@Narice what happens when you do M-x lsp?

**Comment by Narice** [comment]
well, it just says: `LSP :: No LSP server for fundamental-mode(check *lsp-log*).`
and there is no lsp-log buffer I can change to.
Do you think it might be linked to doom emacs instead of lsp-mode directly?

**Comment by yyoncho** [comment]
@Narice you should have that buffer in `gdscript-mode` not in fundamental-mode.

**Comment by NathanLovato** [comment]
It's working out of the box here, but you need [gdscript-mode](https://github.com/GDQuest/emacs-gdscript-mode) installed and configured to activate when opening gdscript files. Ensure that lsp is up to date as well.

**Comment by Narice** [comment]
oh! I didn't think installing gdscript-mode was necessary
thank you very much!
Have a nice day and stay safe ;)

**Comment by dmille** [comment]
I've been trying to get lsp working with gdscript. I've compiled godot 3.2 with the cherrypick mentioned here https://github.com/godotengine/godot/pull/35864#issuecomment-602099845. And I'm running `emacs -q` with the minimal lsp environment from https://github.com/emacs-lsp/lsp-mode/blob/master/scripts/lsp-start-plain.el (with gdscript added to the pkg-list).

This is the error I keep getting when I try and save my .gd file:

`Error processing message (error "Stack overflow in equal").
Error processing message (lsp-unknown-message-type #s(hash-table size 3 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("error" #s(hash-table size 2 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("code" -32601 "message" "Method not found")) "id" nil "jsonrpc" "2.0"))). [2 times]`

And I also get this error everytime lsp connects to the language server for the first time:
`Warning (lsp-mode): Unknown notification: gdscript/capabilities`

If anyone has any ideas, or if there's anything else I could provide, please let me know.

**Comment by ofrank123** [comment]
Yes this is an issue with the current language server implementation on Godot's end.  I'll try to fix it as I am the one who ported the language server to TCP.

**Pull request merged**